xnmt/self_attention.py

```diff
@@ -0,0 +1,87 @@
+import numpy as np
+import dynet as dy
+from math import sqrt
+
+from xnmt.events import register_xnmt_handler, handle_xnmt_event
+from xnmt.expression_sequence import ExpressionSequence, ReversedExpressionSequence
+from xnmt.param_collection import ParamManager
+from xnmt.param_init import GlorotInitializer, ZeroInitializer
+from xnmt.persistence import serializable_init, Serializable, bare, Ref
+from xnmt.transducer import SeqTransducer
+
+class MultiHeadAttentionSeqTransducer(SeqTransducer, Serializable):
+  """
+  This implements the Multi-headed attention layer of "Attention is All You Need":
+  https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf
+
+  Args:
+
+  """
+  yaml_tag = '!MultiHeadAttentionSeqTransducer'
```

Comment:
To support referencing from config files, the module would need to be imported under xnmt/__init__.py